spring-cloud-kubernetes-core/src/main/java/org/springframework/cloud/kubernetes/condition/OnKubernetesConditionSupport.java

```diff
@@ -0,0 +1,50 @@
+package org.springframework.cloud.kubernetes.condition;
+
+import static org.springframework.cloud.kubernetes.Constants.SPRING_CLOUD_KUBERNETES_CLIENT_PROPS;
+
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.ConfigBuilder;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import org.springframework.core.env.PropertyResolver;
+
+final class OnKubernetesConditionSupport {
+
+	private OnKubernetesConditionSupport() {
+
+	}
+
+	// We need to create a KubernetesClient ourselves
+	// since the KubernetesClient Bean is not available when the Condition is evaluated
+	static KubernetesClient getKubernetesClient(PropertyResolver propertyResolver) {
+		Config base = Config.autoConfigure(null);
+		Config properties = new ConfigBuilder(base)
+			.withMasterUrl(propertyResolver.getProperty(
+				SPRING_CLOUD_KUBERNETES_CLIENT_PROPS + ".master-url",
+				base.getMasterUrl()
+				)
+			)
+			.withCaCertFile(propertyResolver.getProperty(
+				SPRING_CLOUD_KUBERNETES_CLIENT_PROPS + ".ca-cert-file",
+				base.getCaCertFile()
+				)
+			)
+			.withCaCertData(propertyResolver.getProperty(
+				SPRING_CLOUD_KUBERNETES_CLIENT_PROPS + ".ca-cert-data",
+				base.getCaCertData()
+				)
+			)
+			.build();
+
+		return new DefaultKubernetesClient(properties);
+	}
+
+	static boolean isServerUp(KubernetesClient kubernetesClient) {
+		try {
+			kubernetesClient.rootPaths();
```

Comment:
Can you explain a bit more what's going on here? Generally, conditions evaluate something in the environment or on the classpath. I'm hesitant to have something that tries to execute against a remote api to determine the condition.

Reply:
This is needed because the whole point of `ConditionalOnKubernetes` is to ensure that a bean is registered when a Kubernetes server is accessible. 

The simplest scenario that I envision is that it can be used to differentiate the beans registered between environments that the code runs in.

Another reason that this is needed is that if this annotation is going to be extended to support matching on the server version (which I plan to do once the relevant functionality is added to `KubernetesClient`), then a connection to the server will be needed anyway.

Reply:
I'm afraid I've never seen a boot condition used that way. Boot doesn't test a database before we create a datasource bean. This is also a potentially very bad for performance. I'd like to see if there's a different way to go about this.

Reply:
On cloudfoundry, we key off of a spring profile named 'cloud'. We don't hit CF api endpoints.

Reply:
That seems strange to me as well. @wilkinsona or @philwebb, thoughts on executing network operation on a conditional?

Reply:
I don’t have all the context, but I don’t like the sound of making a network call in a condition as it’ll hurt performance. Perhaps there’s something that can be down with environment variables instead? That’s how Boot’s [`CloudPlatform`](https://github.com/spring-projects/spring-boot/blob/d3c34ee3d1bfd3db4a98678c524e145ef9bca51c/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/cloud/CloudPlatform.java) is implemented

Reply:
Thank you very much for the input. 

The reasoning behind the network call is to provide conditions based on some data from the server (for example, the server version). 

I totally understand the performance concern, which is why I suggested that it could be controlled by a property - when the property is set, then the regular functionality is enabled, otherwise it's not.

Reply:
I see two issues, with environment variables:

- not sure if existing environment variables cover our needs.
- at dev time the user will have to toy with those.

From my point of view its preferable to perform the actual call and provide the functionality in a transparent way at the cost of boot speed (as long as its well documented).

Note, that kubernetes is no database. In fact its unlike any other piece of software out there in that it has frequent releases that all introduce new features all the time. This increases the need to allow users to craft version specific config/beans etc.